syntax = "proto2";

package prga.common;

message BitConnection {
  message ConnectionAction {
    extensions 1 to max;
  }

  required string input = 1;
  optional ConnectionAction action = 2;
}

message PortBit {
  repeated BitConnection connections = 1;
}

message Port {
  required string name = 1;
  repeated PortBit bits = 2;
}

message Mode {
  message ModeAction {
    extensions 1 to max;
  }

  required string name = 1;
  optional ModeAction action = 2;
}

message Instance {
  enum Type {
    LUT = 1;
    MULTIMODE = 2;  // for generic multi-mode instance including inout
    CUSTOM = 3;     // for generic model instance
  }

  message InstanceAction {
    extensions 1 to max;
  }

  required string name = 1;
  repeated Port ports = 2;
  repeated Mode modes = 3;
  optional InstanceAction action = 4;
  required Type type = 5;
}

message Block {
  message BlockAction {
    extensions 1 to max;
  }

  required string name = 1;
  repeated Port ports = 2;
  optional BlockAction action = 3;
  repeated Instance instances = 4;
}

message Placement {
  message PlacementAction {
    extensions 1 to max;
  }

  required uint32 x = 1;
  required uint32 y = 2;
  optional uint32 subblock = 3 [default = 0];
  optional PlacementAction action = 4;
}

message RoutingEdge {
  message RoutingAction {
    extensions 1 to max;
  }

  required uint64 src = 1;
  required uint64 sink = 2;
  optional RoutingAction action = 3;
}

message Packet {
  message Attachment {
    extensions 1 to max;
  }

  repeated Block blocks = 1;
  repeated Placement placements = 2;
  repeated RoutingEdge edges = 3;
  repeated Attachment attachments = 4;
}

message Header {
  message HeaderAction {
    extensions 1 to max;
  }

  required uint64 signature = 1;  // a signature specific to a config circuitry
                                  // type
  required uint32 width = 2;
  required uint32 height = 3;
  required uint64 node_size = 4;  // total number of routing nodes
  optional HeaderAction action = 5;
}
